x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java

```diff
@@ -0,0 +1,273 @@
+                        try {
+                            // cache file was evicted during the range fetching, read bytes directly from source
+                            bytesRead += copySource(pos, pos + len, buffer, off);
+                            continue;
+                        } catch (Exception inner) {
+                            e.addSuppressed(inner);
+                        }
+                    }
+                    throw new IOException("Fail to read data from cache", e);
+
+                }
+            }
+            assert bytesRead == length : "partial read operation, read [" + bytesRead + "] bytes of [" + length + "]";
+            assert parent == null || cacheFile.get() == null;
+        }
+
+        int readCacheFile(FileChannel fc, long end, long position, byte[] buffer, int offset, long length) throws IOException {
+            assert assertFileChannelOpen(fc);
+            return Channels.readFromFileChannel(fc, position, buffer, offset, Math.toIntExact(Math.min(length, end - position)));
+        }
+
+        @SuppressForbidden(reason = "Use positional writes on purpose")
+        void writeCacheFile(FileChannel fc, long start, long end) throws IOException {
+            assert assertFileChannelOpen(fc);
+            final byte[] copyBuffer = new byte[Math.toIntExact(Math.min(8192L, end - start))];
+            try (IndexInput input = in.openInput(fileName, ioContext)) {
+                if (start > 0) {
+                    input.seek(start);
+                }
+                int bytesCopied = 0;
+                long remaining = end - start;
+                while (remaining > 0) {
+                    final int size = (remaining < copyBuffer.length) ? Math.toIntExact(remaining) : copyBuffer.length;
+                    input.readBytes(copyBuffer, 0, size);
+                    fc.write(ByteBuffer.wrap(copyBuffer, 0, size), start + bytesCopied);
+                    bytesCopied += size;
+                    remaining -= size;
+                }
+            }
+        }
+
+        @Override
+        protected void seekInternal(long pos) throws IOException {
+            if (pos > length()) {
+                throw new EOFException("Reading past end of file [position=" + pos + ", length=" + length() + "] for " + toString());
+            } else if (pos < 0L) {
+                throw new IOException("Seeking to negative position [" + pos + "] for " + toString());
+            }
+        }
+
+        @Override
+        public CacheBufferedIndexInput clone() {
+            final CacheBufferedIndexInput clone = (CacheBufferedIndexInput) super.clone();
+            clone.cacheFile = new AtomicReference<>();
+            clone.closed = new AtomicBoolean(false);
+            clone.parent = (this.parent != null ? this.parent : this);
+            assert clone.parent.parent == null : "parent must not be a clone";
+            return clone;
```

Comment:
I think this means that clones will share locks with their parent, but slices do not. Doesn't seem to be a problem, but it is inconsistent.

Reply:
You're right, cloning object might be trappy sometimes. Since we moved to `synchronized` blocks it is not an issue anymore but thanks for pointing this.